nes-network/network/src/sender/channel.rs

```diff
@@ -0,0 +1,387 @@
+            channel_identifier,
+        ))
+        .await?;
+
+    let response = read
+        .next()
+        .await
+        .ok_or("Connection closed during identification")??;
+
+    let stream = Channel {
+        writer: write.into_inner().into_inner(),
+        reader: read.into_inner().into_inner(),
+    };
+
+    match response {
+        IdentificationResponse::Ok => Ok(data_channel_sender(stream)),
+        IdentificationResponse::NotOk => {
+            Err("Could not establish connection. Identification was denied".into())
+        }
+    }
+}
+
+const MAX_PENDING_ACKS: usize = 64;
+
+pub(super) enum ChannelControlMessage {
+    Data(TupleBuffer),
+    Flush(oneshot::Sender<bool>),
+}
+pub(super) type ChannelControlQueue = async_channel::Sender<ChannelControlMessage>;
+pub(super) type ChannelControlQueueListener = async_channel::Receiver<ChannelControlMessage>;
+pub(super) struct ChannelHandler<R: AsyncRead + Unpin, W: AsyncWrite + Unpin> {
+    cancellation_token: CancellationToken,
+    pending_writes: VecDeque<TupleBuffer>,
+    wait_for_ack: HashMap<OriginSequenceNumber, TupleBuffer>,
+    writer: DataChannelSenderWriter<W>,
+    reader: DataChannelSenderReader<R>,
+    queue: ChannelControlQueueListener,
+}
+
+enum ErrorOrStatus {
+    Error(Error),
+    Status(ChannelHandlerStatus),
+}
+type InternalResult<T> = std::result::Result<T, ErrorOrStatus>;
+
+impl<R: AsyncRead + Unpin, W: AsyncWrite + Unpin> ChannelHandler<R, W> {
+    pub fn new(
+        cancellation_token: CancellationToken,
+        queue: ChannelControlQueueListener,
+        reader: DataChannelSenderReader<R>,
+        writer: DataChannelSenderWriter<W>,
+    ) -> Self {
+        Self {
+            cancellation_token,
+            pending_writes: Default::default(),
+            wait_for_ack: Default::default(),
+            reader,
+            writer,
+            queue,
+        }
+    }
+    async fn handle_request(
+        &mut self,
+        channel_control_message: ChannelControlMessage,
+    ) -> InternalResult<()> {
+        match channel_control_message {
+            ChannelControlMessage::Data(data) => self.pending_writes.push_back(data),
+            ChannelControlMessage::Flush(done) => {
+                Self::cancelable(&self.cancellation_token, self.writer.flush())
+                    .await?
+                    .map_err(|e| ErrorOrStatus::Error(e.into()))?;
+                let _ = done.send(self.pending_writes.is_empty() && self.wait_for_ack.is_empty());
+            }
+        }
+        Ok(())
+    }
+    fn handle_response(&mut self, response: DataChannelResponse) -> InternalResult<()> {
+        match response {
+            DataChannelResponse::Close => {
+                info!("Channel Closed by other receiver");
+                return Err(ErrorOrStatus::Status(
+                    ChannelHandlerStatus::ClosedByOtherSide,
+                ));
+            }
+            DataChannelResponse::NAckData(seq) => {
```

Comment:
I would remove it when it is currently not used and therefore also not tested.

Reply:
I have tests in a future pr, that only use the datachannel. They will use the nack.